monix-eval/shared/src/main/scala/monix/eval/TaskLocal.scala

```diff
@@ -134,7 +134,7 @@ final class TaskLocal[A] private (default: => A) {
     Task.suspend {
       val saved = ref.value
       ref.update(value)
-      task.doOnFinish(_ => restore(saved))
+      task.bracket(r => Task.now(r))(_ => restore(saved))
```

Comment:
If you want to handle cancellation, `task` has to be used inside bracket. E.g.

```scala
Task.unit.bracket(_ => task)(_ => restore(saved))
```

Also `bindL` needs to have similar change, so maybe it's easier to implement `bind` as `bindL(Task.now(value))`